File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/HarnessCheckRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;

/**
 * Strategy interface implemented by each harness validation rule.
 */
public interface HarnessCheckRule {
    /**
     * Determines whether this rule applies to the given manifest.
     *
     * @param manifest the manifest JSON node
     * @return true if this rule should be validated
     */
    boolean applies(JsonNode manifest);

    /**
     * Validates the given project directory against this rule.
     *
     * @param projectDir the project root directory
     * @param manifest the manifest JSON node
     * @return a list of findings; empty if validation passes
     * @throws MojoExecutionException if validation fails
     */
    List<Finding> validate(Path projectDir, JsonNode manifest) throws MojoExecutionException;
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/HarnessCheckRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path

/**
 * Interface for harness validation rules.
 */
interface HarnessCheckRule {
	/**
	 * Check whether this rule applies based on manifest configuration.
	 *
	 * @param manifest The harness validation manifest.
	 * @return true if the rule is enabled, false otherwise.
	 */
	fun applies(manifest: JsonObject): Boolean

	/**
	 * Validate and return a list of findings.
	 *
	 * @param manifest The harness validation manifest.
	 * @param root The root path of the project.
	 * @param psiResults PSI analysis results, may be null.
	 * @return List of findings; empty list if no issues found.
	 */
	fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults? = null): List<Finding>
}
